### ISSUE TITLE
Synonyms to a column cannot be used in load_only

### ISSUE THREAD

**Issue by jace** [sqlalchemy/sqlalchemy#6272]
**Describe the bug**
Synonyms to a column cannot be used in a query with `options(load_only(synonym_name))`. This worked in 1.3 but fails in 1.4.

**Expected behavior**

The query should work using the column behind the synonym.

**To Reproduce**

With pytest:

```py
from sqlalchemy import Column, Integer, String, create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import load_only, sessionmaker, synonym

import pytest

engine = create_engine('sqlite:///', future=True)
Base = declarative_base()
Base.metadata.bind = engine


class TestModel(Base):
    __tablename__ = 'test_model'

    id = Column(Integer, primary_key=True)  # NOQA: A003
    direct = Column(String)
    alias = synonym('direct')


@pytest.fixture
def session():
    Base.metadata.create_all()
    Session = sessionmaker(engine, future=True)
    with Session() as session:
        yield session


def test_query_direct(session):
    """This works."""
    session.query(TestModel).filter(TestModel.direct == "x").options(
        load_only(TestModel.direct)
    ).all()


def test_query_alias(session):
    """This fails."""
    session.query(TestModel).filter(TestModel.alias == "x").options(
        load_only(TestModel.alias)
    ).all()
```

**AttributeError: Neither 'property' object nor 'Comparator' object associated with TestModel.alias has an attribute '_extra_criteria'**

```
../../lib/python3.8/site-packages/sqlalchemy/orm/query.py:2685: in all
    return self._iter().all()
../../lib/python3.8/site-packages/sqlalchemy/orm/query.py:2820: in _iter
    result = self.session.execute(
../../lib/python3.8/site-packages/sqlalchemy/orm/session.py:1670: in execute
    result = conn._execute_20(statement, params or {}, execution_options)
../../lib/python3.8/site-packages/sqlalchemy/engine/base.py:1520: in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
../../lib/python3.8/site-packages/sqlalchemy/sql/elements.py:313: in _execute_on_connection
    return connection._execute_clauseelement(
../../lib/python3.8/site-packages/sqlalchemy/engine/base.py:1381: in _execute_clauseelement
    compiled_sql, extracted_params, cache_hit = elem._compile_w_cache(
../../lib/python3.8/site-packages/sqlalchemy/sql/elements.py:548: in _compile_w_cache
    compiled_sql = self._compiler(
../../lib/python3.8/site-packages/sqlalchemy/sql/elements.py:570: in _compiler
    return dialect.statement_compiler(dialect, self, **kw)
../../lib/python3.8/site-packages/sqlalchemy/sql/compiler.py:766: in __init__
    Compiled.__init__(self, dialect, statement, **kwargs)
../../lib/python3.8/site-packages/sqlalchemy/sql/compiler.py:455: in __init__
    self.string = self.process(self.statement, **compile_kwargs)
../../lib/python3.8/site-packages/sqlalchemy/sql/compiler.py:490: in process
    return obj._compiler_dispatch(self, **kwargs)
../../lib/python3.8/site-packages/sqlalchemy/sql/visitors.py:81: in _compiler_dispatch
    return meth(self, **kw)
../../lib/python3.8/site-packages/sqlalchemy/sql/compiler.py:2966: in visit_select
    compile_state = select_stmt._compile_state_factory(
../../lib/python3.8/site-packages/sqlalchemy/sql/base.py:492: in create_for_statement
    return klass.create_for_statement(statement, compiler, **kw)
../../lib/python3.8/site-packages/sqlalchemy/orm/context.py:579: in create_for_statement
    opt.process_compile_state(self)
../../lib/python3.8/site-packages/sqlalchemy/orm/strategy_options.py:182: in process_compile_state
    self._process(compile_state, not bool(compile_state.current_path))
../../lib/python3.8/site-packages/sqlalchemy/orm/strategy_options.py:705: in _process
    val._bind_loader(
../../lib/python3.8/site-packages/sqlalchemy/orm/strategy_options.py:889: in _bind_loader
    if not loader._generate_path(
../../lib/python3.8/site-packages/sqlalchemy/orm/strategy_options.py:328: in _generate_path
    if attr._extra_criteria:
../../lib/python3.8/site-packages/sqlalchemy/orm/attributes.py:588: in __getattr__
    util.raise_(
```

**Versions.**
 - OS: Linux x86_64
 - Python: 3.8.5
 - SQLAlchemy: 1.4.7
 - Database: NA
 - DBAPI: NA

**Comment by zzzeek** [comment]
i dont have a quick fix for this one, I likely have to put 1.4.8 out with what i have wiating to go out

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **master** branch:

**synonym fixes and enhancements** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2749
